Introduced CellDataTransfer.

marcfehling opened the pull request:
This PR introduces the `CellDataTransfer` class for non-distributed triangulations. The distributed pendant has been introduced in #7537.

Similarly to the `SolutionTransfer` class, we store `active_cell_indices` on the unrefined mesh and use them to transfer data from the initial to the refined mesh.

We use the `CoarseningStrategy` struct to decide how to handle data during coarsening. However for `CellDataTransfer`, we need the values on all coarsened cells not before, but after coarsening, when the cells no longer exist. Thus, some changes were necessary.

For more flexibility, we moved the `CoarseningStrategy` struct out of the scope of the `p::d::CellDataTransfer` class. It is now an independent struct in the `dealii` namespace defined in `deal.II/numerics/cell_data_transfer.h`. Shall we introduce a separate header for this struct, e.g. `deal.II/numerics/coarsening_strategy.h`?
EDIT: Moved it to `grid/grid_tools.h` now.

Requested by #7515.

marcfehling commented:
Do we need a `DEPRECATED` version of the old `CoarseningStrategies` struct?

marcfehling commented:
ping

jppelteret commented:
Hi @marcfehling. Sorry that its taken so long for one of us to respond. It seems like we're all preoccupied with other things at the moment -- not ideal at all, especially given that we're really interested in, and grateful for, your contributions. I think that this is cool, and this sort of mechanisms to transfer data would be helpful to facilitate that transfer of data in the `CellDataStorage` class, which I believe might only be possible in the distributed case right now. Thanks a lot!

> It is now an independent struct in the dealii namespace defined in deal.II/numerics/cell_data_transfer.h. Shall we introduce a separate header for this struct, e.g. deal.II/numerics/coarsening_strategy.h?

Since `CoarseningStrategy` is referenced in both `deal.II/distributed/cell_data_transfer.h` and `deal.II/numerics/cell_data_transfer.h `, I think that putting it in a some common header is a good idea. I see that you've now put it in `GridTools`. That's quite a big header to drag into `cell_data_transfer.h`, so I think that your original idea of putting it in `deal.II/numerics/coarsening_strategy.h` might be best. Since you've also put it in its own `doxygen` section, its clear that its intended use is quite context specific.

> Do we need a DEPRECATED version of the old CoarseningStrategies struct and the old p::d::CellDataTransfer constructor?
> EDIT: Flagging parts of this class as DEPRECATED is not that easy: We store an std::function which takes a different set of parameters with this PR. Thus, we need to deprecate the old pointer and introduce the new pointer at the same time. And the callback function would have to call the corresponding one which is not a nullptr.
> Shall we follow this approach? Or shall we accept this incompatibility?

Hmm... its always best for us to deprecate functionality if its at all possible. But this does seem like a bit of a difficult case to handle. If I understand correctly, you're taking about this function:
OLD:
```c++
const std::function<typename VectorType::value_type(
  const typename parallel::distributed::Triangulation<dim, spacedim>::
  cell_iterator & parent,
  const VectorType &input_vector)>
coarsening_strategy;
```
NEW:
```c++
const std::function<value_type(
  const std::vector<value_type> &children_values)>
coarsening_strategy;
```
We could keep and deprecate the old `CoarseningStrategy` struct in the distributed `CellDataTransfer`, and in the (old/deprecated) constructor just translate its meaning by forwarding it on to the new constructor. Something like
```c++
CellDataTransfer(
  const parallel::distributed::Triangulation<dim, spacedim> &triangulation,
  const bool transfer_variable_size_data = false
  const std::function<typename VectorType::value_type(
    const typename parallel::distributed::Triangulation<dim, spacedim>::
    cell_iterator & parent,
    const VectorType &input_vector)>
  coarsening_strategy = &CoarseningStrategies::check_equality)
: CellDataTransfer (
  triangulation, transfer_variable_size_data,
  (coarsening_strategy == &CoarseningStrategies::check_equality 
    ? &GridTools::CoarseningStrategies<value_type>::check_equality 
    : ...<check all options; maybe do with with some internal helper struct>... )
)
```
Do you think that this could work?

drwells commented:
We have a ton of open PRs now and I tagged myself with the intent of reviewing this early this week. @jppelteret are you willing to take this over?

jppelteret commented:
Sorry @drwells, I wish I could but I'm not able say with certainty when my next free chunk of time will be :-( I had a bit of a unicorn moment today when I had some time to devote to browse through the open PRs, and so I added my 2c here! If the opportunity arises then I will come back and look at this, but (if possible) that would likely only be next weekend. But I'd certainly appreciate it if you're able to give it some scrutiny when you have a free moment.

drwells commented:
I will review this later today; @jppelteret's notes are a good start.

marcfehling commented:
Moved `CoarseningStrategies` to a separate header.
Thank you @jppelteret for the hints on deprecating the constructor. The implementation turned out to be a bit whacky, but serves its purpose well I guess. However I could not find a way to make this work for general functions with the old set of parameters and will throw an exception instead.

drwells commented on include/deal.II/numerics/cell_data_transfer.h:
```diff
@@ -0,0 +1,212 @@
+ * distributed_data_to_transfer (mpi_communicator,
+ *                               triangulation.n_active_cells(),
+ *                               triangulation.n_locally_owned_active_cells());
+ * for (const auto cell& : triangulation.active_cell_iterators ())
```
```suggestion
 * for (const auto &cell : triangulation.active_cell_iterators())
```

drwells commented on include/deal.II/numerics/cell_data_transfer.h:
```diff
@@ -0,0 +1,212 @@
+   * @param[in] transfer_variable_size_data Specify whether your VectorType
+   *   container stores values that differ in size. A varying amount of data
+   *   may be packed per cell, if for example the underlying ValueType of
+   *   the VectorType container is a container itself.
```
It doesn't look like `transfer_variable_size_data` is in this constructor. Should it be?

marcfehling replied:
Woops, I blindly copied that from the `p::d::CellDataTransfer` class. Thanks for noticing!

drwells commented on include/deal.II/numerics/cell_data_transfer.h:
```diff
@@ -0,0 +1,212 @@
+   * validate user inputs after refinement happened.
+   */
+  unsigned int n_active_cells_pre;
+#endif
```
I think it would be better to just write
```cpp
(void)n_active_cells_pre;
```
instead of changing the class members in different modes.

drwells commented on include/deal.II/numerics/cell_data_transfer.templates.h:
```diff
@@ -0,0 +1,196 @@
+  {
+    // Do nothing for std::vector as VectorType.
+  }
+} // namespace
```
Having anonymous namespaces in headers is a bit risky: I recommend putting these in `namespace internal` instead.

drwells commented on include/deal.II/numerics/coarsening_strategies.h:
```diff
@@ -0,0 +1,94 @@
+ * @author Marc Fehling, 2019
+ */
+template <typename value_type>
+struct CoarseningStrategies
```
Would this work if we used a namespace instead, or would that screw up taking addresses of functions?

marcfehling replied:
I just tried it out: Works like a charm!

drwells commented on include/deal.II/numerics/coarsening_strategies.h:
```diff
@@ -0,0 +1,94 @@
+ *
+ * @author Marc Fehling, 2019
+ */
+template <typename value_type>
```
Do we have some template type check we could use here to ensure that `median` would work correctly?

marcfehling replied:
Were you referring to the `mean` function? I guess the code wouldn't compile if `value_type` has no zero value and no `operator+=`.

Otherwise I don't know what you mean by `median`. Would you elaborate on that?

marcfehling replied:
Are you suggesting a new `median` function that returns the median of all child values? (The number of children is always even)

drwells replied:
Sorry, I meant `mean`.

marcfehling replied:
We may use the  concepts feature that has been introduced in [C++20](https://en.cppreference.com/w/cpp/language/constraints) or [Boost 1.4](https://www.boost.org/doc/libs/1_70_0/libs/concept_check/concept_check.htm) for this.

But I think using [type traits](https://en.cppreference.com/w/cpp/types) to check for arithmetic types should be sufficient for now until someone demands a more versatile implementation. But I don't know how to sum/mean `bool` variables - we may just exclude them.

drwells commented on include/deal.II/numerics/coarsening_strategies.h:
```diff
@@ -0,0 +1,94 @@
+ * how to handle data of child cells which will be coarsened. Or in other
+ * words, which data should be stored in the corresponding parent cell.
+ *
+ * In this namespace, we offer a few strategies that cope with this
```
```suggestion
 * In this struct, we offer a few strategies that cope with this
```

drwells commented on include/deal.II/numerics/coarsening_strategies.h:
```diff
@@ -0,0 +1,94 @@
+               "Values on cells that will be coarsened are not equal!"));
+
+    return value;
+  }
```
It doesn't make a big difference, but we usually separate function definitions and declarations (even if they are in the same file). We should do this.

drwells commented on include/deal.II/numerics/cell_data_transfer.templates.h:
```diff
@@ -0,0 +1,196 @@
+         ++child_index)
+      {
+        const auto child = refined.first->child(child_index);
+        Assert(child->active(), ExcInternalError());
```
Is this always true? Couldn't I do the following:
```
+-----------+
|           |
|           |
|           |
|           |
|           |
+-----------+
```
to
```
+--+--+-----+
|__|__|     |
|  |  |     |
+--+--+-----+
|     |     |
|     |     |
+-----+-----+
```
then a refined cell could have grandchildren.

marcfehling replied:
This is indeed possible.

However, `Triangulation::prepare_coarsening_and_refinement()` cleans up all coarsening flags:  We only keep these flags whenever all siblings are active and flagged for coarsening as well. We require this specific function to be called beforehand.

This means that the exception thrown should not be `ExcInternalError`, but a custom one giving users a hint towards the `prepare()` function. Thank you for pointing that out!

drwells commented on include/deal.II/numerics/cell_data_transfer.templates.h:
```diff
@@ -0,0 +1,196 @@
+                   ++child_index)
+                {
+                  const auto sibling = parent->child(child_index);
+                  Assert(sibling->active(), ExcInternalError());
```
Like my refinement example: could we coarsen more than one level at once? I think we can but perhaps there is some other precondition enforced that I don't know about.

drwells commented on include/deal.II/numerics/cell_data_transfer.templates.h:
```diff
@@ -0,0 +1,196 @@
+         ExcDimensionMismatch(in.size(), n_active_cells_pre));
+  Assert(out.size() == triangulation->n_active_cells(),
+         ExcDimensionMismatch(out.size(), triangulation->n_active_cells()));
+#endif
```
```suggestion
#else
  (void)n_active_cells_pre;
#endif
```

marcfehling commented:
@drwells, thank your for your extensive review! Addressed all suggestions but one. Would you give me more information on the remaining one.

I will squash all commits into one once you approve this PR. I cleaned up some other code on the last commit since there were exceptions that recur. Let me know what you think of it!

marcfehling commented on include/deal.II/distributed/cell_data_transfer.templates.h:
```diff
@@ -247,10 +300,26 @@ namespace parallel
+                      const auto child = cell->child(child_index);
+                      Assert(child->active() && child->coarsen_flag_set(),
+                             typename dealii::Triangulation<
+                               dim>::ExcObscureCoarseningFlags());
```
Do you think placing this particular one in `Triangulation<dim>` is reasonable?

marcfehling commented on source/distributed/cell_weights.cc:
```diff
@@ -145,18 +145,23 @@ namespace parallel
-                     "that is dominated by all children of a cell you are "
-                     "trying to coarsen!"));
+                   typename dealii::hp::FECollection<
+                     dim>::ExcNoDominatedFiniteElementAmongstChildren());
```
Do you think placing this particular one in `hp::FECollection<dim>` is reasonable?

drwells commented:
Yes, the exceptions look good, but I think the name `ExcObscureCoarseningFlags` is misleading; the flags are not obscure (though this error surely is :)), they are inconsistent. How about `ExcInconsistentCoarseningFlags`?

Otherwise :+1:

marcfehling commented:
Did some cleanup and squashed the commits into four separate ones. Addressed all suggestions.

Ready for final review and/or testing!

drwells commented:
/rebuild
